### PR TITLE
RT-2.10: IS-IS  change LSP lifetime

### PR DESCRIPTION
**Pull request by divyarema**
Initial commit for RT2.10
It resolves https://github.com/openconfig/featureprofiles/issues/1619

### CODE REVIEW

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+## RT-2.10: IS-IS change LSP lifetime
+
+## Summary
+
+* Changing the lsp lifetime and verifying isis lsp parametrs
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md:5]: *parameters

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+## RT-2.10: IS-IS change LSP lifetime
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md:1]: should only start with a single #

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+## Summary
+
+* Changing the lsp lifetime and verifying isis lsp parametrs
+
+## Procedure
+
+* TestChangeLspLifetime
+
+    * Configure IS-IS for ATE port-1 and DUT port-1.
+    * Modify the default life time of the LSP PDU.
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md:12]: the text alternates between lifetime as one word or two. Standardize to one word

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+const (
+	plenIPv4       = 30
+	plenIPv6       = 126
+	ISISInstance   = "DEFAULT"
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:40]: lowercase this constant name, on need to export

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+	dutAreaAddress = "49.0001"
+	ateAreaAddress = "49.0002"
+	dutSysID       = "1920.0000.2001"
+	lspLifetime    = 500
+	v4Route        = "203.0.113.0/30"
+	v6Route        = "2001:db8::203:0:113:0/126"
+	v4Ip           = "203.0.113.1"
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:47]: go initialism style says IP should be capitalized, i.e. "v4IP" and "v6IP"

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+		})
+		t.Run("Traffic checks", func(t *testing.T) {
+			ate.Traffic().Start(t, flows...)
+			time.Sleep(time.Second * 15)
+			ate.Traffic().Stop(t)
+
+			for _, flow := range flows {
+				t.Log("Checking flow telemetry...")
+				telem := gnmi.OC()
+				Loss := gnmi.Get(t, ate, telem.Flow(flow.Name()).LossPct().State())
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:309]: lowercase the local variable

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md (hunk 1 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md]: taken care in latest commit

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md (hunk 3 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md]: addressed in latest commit

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 4 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go]: addressed in latest commit

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 5 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go]: addressed in latest commit

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 6 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go]: addressed in latest commit

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md (hunk 2 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md]: modified in recent commit

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+	gnmi.Replace(t, dut, dc.Interface(i2.GetName()).Config(), i2)
+
+	if deviations.ExplicitPortSpeed(dut) {
+		fptest.SetPortSpeed(t, dut.Port(t, "port1"))
+		fptest.SetPortSpeed(t, dut.Port(t, "port2"))
+	}
+}
+
+// configureIsisDut configures isis on DUT.
+func configureIsisDut(t *testing.T, dut *ondatra.DUTDevice, intfName string, dutAreaAddress, dutSysID string) {
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:103]: go style says "Isis" should be stylized "ISIS" and "Dut" as "DUT," as you already did above. So technically this should be configureISISDUT, but you could make it configureISIS for short.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+	v6Flow := ate.Traffic().NewFlow("v6Flow").
+		WithSrcEndpoints(srcIntf).WithDstEndpoints(dstIntf).
+		WithHeaders(ondatra.NewEthernetHeader(), v6Header).WithFrameRateFPS(100)
+
+	return []*ondatra.Flow{v4Flow, v6Flow}
+}
+
+// TestIsisChangeLspLifetime verifies isis lsp telemetry paramters with configured lsp lifetime.
+func TestIsisChangeLspLifetime(t *testing.T) {
```
Comment (greg-dennis) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:194]: "Isis" -> "ISIS"
"Lsp" -> "LSP"

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 7 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go]: addressed in latest commit

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 8 above)
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go]: addressed in latest commit

**Review by greg-dennis — APPROVED** [review]
If all the presubmit checks pass, approved to merge

**Comment by greg-dennis** [comment]
@wenovus can provide direction to resolving the OC path static analysis failure

**Review by greg-dennis — CHANGES_REQUESTED** [review]
Un-approving until the static analysis failure is fixed.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+			if got := gnmi.Get(t, dut, statePath.Level(2).Lsp(ateLspID).Tlv(oc.IsisLsdbTypes_ISIS_TLV_TYPE_IPV6_REACHABILITY).Ipv6Reachability().Prefix(v6Route).Prefix().State()); got != v6Route {
+				t.Errorf("FAIL- Expected v6 route not found in isis, got %v, want %v", got, v6Route)
+			}
+			if got := gnmi.Get(t, dut, gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Afts().Ipv4Entry(v4Route).Prefix().State()); got != v4Route {
+				t.Errorf("FAIL- Expected v4 route not found in aft, got %v, want %v", got, v4Route)
+			}
+			if got := gnmi.Get(t, dut, gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Afts().Ipv6Entry(v6Route).Prefix().State()); got != v6Route {
+				t.Errorf("FAIL- Expected v6 route not found in aft, got %v, want %v", got, v6Route)
+			}
+		})
```
[the remark this answers is not in the document]
Comment (divyarema) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go]: modified in latest commit

**Review by greg-dennis — APPROVED** [review]
Re-approving for Go quality

**Review by sachendras — APPROVED** [review]

**Review by greg-dennis — APPROVED** [review]

**Review by wenovus — APPROVED** [review]
The IPv4Entry() query lines LGTM. Leaving the review of the testing logic to others.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+	i2Dut := topo.AddInterface(atePort2attr.Name).WithPort(port2)
+	i2Dut.IPv4().WithAddress(atePort2attr.IPv4CIDR()).WithDefaultGateway(dutPort2Attr.IPv4)
+	i2Dut.IPv6().WithAddress(atePort2attr.IPv6CIDR()).WithDefaultGateway(dutPort2Attr.IPv6)
+
+	isisDut := i1Dut.ISIS()
+	isisDut.
+		WithAreaID(ateAreaAddress).
+		WithTERouterID(atePort1attr.IPv4).
+		WithNetworkTypePointToPoint().
+		WithLevelL2()
```
Comment (arvbaska1) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:154]: Please set link metric as well from the ATE using WithMetric()
By default the metric from ATE is 0.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+	// Global configs
+	globalIsis.Net = []string{fmt.Sprintf("%v.%v.00", dutAreaAddress, dutSysID)}
+	globalIsis.GetOrCreateAf(oc.IsisTypes_AFI_TYPE_IPV4, oc.IsisTypes_SAFI_TYPE_UNICAST).Enabled = ygot.Bool(true)
+	globalIsis.GetOrCreateAf(oc.IsisTypes_AFI_TYPE_IPV6, oc.IsisTypes_SAFI_TYPE_UNICAST).Enabled = ygot.Bool(true)
+	globalIsis.LevelCapability = oc.Isis_LevelType_LEVEL_2
+
+	globalIsis.GetOrCreateTimers().LspLifetimeInterval = ygot.Uint16(lspLifetime)
+
+	// Interface configs
+	intf := isis.GetOrCreateInterface(intfName)
```
Comment (arvbaska1) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:125]: Please use the deviation missing_isis_interface_afi_safi_enable here similar to 
https://github.com/openconfig/featureprofiles/blob/main/feature/experimental/isis/ate_tests/internal/session/session.go#L133

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+	d := &oc.Root{}
+	configPath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_ISIS, isisInstance)
+	netInstance := d.GetOrCreateNetworkInstance(deviations.DefaultNetworkInstance(dut))
+	prot := netInstance.GetOrCreateProtocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_ISIS, isisInstance)
+
+	if !deviations.ISISprotocolEnabledNotRequired(dut) {
+		prot.Enabled = ygot.Bool(true)
+	}
+	isis := prot.GetOrCreateIsis()
+	globalIsis := isis.GetOrCreateGlobal()
```
Comment (arvbaska1) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:114]: Please add deviation isis_single_topology_required similar to 
https://github.com/openconfig/featureprofiles/blob/main/feature/experimental/isis/ate_tests/base_adjacencies_test/base_adjacencies_test.go#L507

**Review by sachendras — APPROVED** [review]

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+* Changing the lsp lifetime and verifying isis lsp parameters
+
+## Procedure
+
+    * Configure IS-IS for ATE port-1 and DUT port-1.
+    * Modify the default lifetime of the LSP PDU.
+    * The default lifetime of the LSP PDU is 1200 seconds.
+        This parameter can be updated using the LSP lifetime parameter.
+        LSP lifetime indicates how long the LSP PDU originated by the DUT should remain in the network. 
+        The DUT regenerates the LSP PDU typically ~300 seconds before its expiration.
```
Comment (arvbaska1) [inline feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md:14]: I think here assumption is LSP refresh interval is dynamically calculated from the LSP lifetime ?
Based on implementation , LSP refresh time can also be after a specific interval.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 10 above)
Comment (sachendras) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:154]: @arvbaska1, can you please address these requests in a different pull? We need to merge this one soon.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 11 above)
Comment (sachendras) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:125]: @arvbaska1, can you please address these requests in a different pull? We need to merge this one soon.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 12 above)
Comment (sachendras) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:114]: @arvbaska1, can you please address these requests in a different pull? We need to merge this one soon.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md (hunk 13 above)
Comment (sachendras) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/README.md:14]: @arvbaska1, can you please address these requests in a different pull? We need to merge this one soon.

File: feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go (hunk 10 above)
Comment (arvbaska1) [reply feature/experimental/isis/ate_tests/isis_change_lsp_lifetime_test/isis_change_lsp_lifetime_test.go:154]: Sure, we can address these comments with another PR.

**Pull request merged**